R/ddl.R

```diff
@@ -0,0 +1,193 @@
+#'  Syntatically valid R code to be executed in the shiny session.
+#'  shouldn't be specified when `code` is specified.
+#'
+#' @param code (`character`, `language`)\cr
+#'  Object containing code to be evaluated to load data. Shouldn't be specified when `expr`
+#'  is specified.
+#'
+#'
+#' @param ui (`shiny.tag`)\cr
+#'   `shiny` ui module containing inputs which `id` correspond to the
+#'   args in the `code`.
+#'
+#' @param server (`function`)\cr
+#'   `shiny` server module returning data. This server should execute
+#'   `code` and return a reactive data containing necessary data. To handle
+#'   evaluation and code masking process it is recommended to use `ddl_run`.
+#'   Package provides universal `username_password_server` which
+#'   runs [ddl_run] function, which returns `teal_data` object.
+#'   Details in the the example
+#'
+#' @param input_mask (`list` named)\cr
+#'   arguments to be substituted in the `code`. These
+#'   argument are going to replace arguments set through
+#'   `ui` and `server`. Example use case is when app user
+#'   is asked to input a password and we'd like to skip this
+#'   input in the reproducible code. Typically users password
+#'   is substituted with `askpass::askpass()` call, so the
+#'   returned code is still executable but secure.
+#'
+#' @param datanames (`character`)\cr
+#'   Names of the objects to be created from the code evaluation.
+#'   If not specified (`character(0)`), all objects will be used to `teal_data` function
+#'   (via `env_list` in `postprocess_fun`).
+#'
+#' @inheritParams teal.data::teal_data
+#'
+#'
+#' @export
+ddl <- function(expr,
+                code,
+                ui = submit_button_ui,
+                input_mask = list(),
+                server = submit_button_server,
+                join_keys = teal.data::join_keys(),
+                datanames) {
+  if (!missing(expr) && !missing(code)) {
+    stop("Only one of `expr` or `code` should be specified")
+  }
+  if (!missing(expr)) {
+    code <- substitute(expr)
+  }
+  if (is.character(code)) {
+    code <- parse(text = code)
+  }
+
+  if (missing(datanames)) {
+    stop("`dataname` argument is required")
+  }
+
+
+  # function creates  object from the code, input and input_mask
+  # function defined here to have access to the arguments
+  ddl_run <- function(input = list()) {
+    checkmate::assert_list(input)
+    if (inherits(input, "reactivevalues")) {
+      input <- shiny::reactiveValuesToList(input)
+    }
+    env <- list2env(list(input = input))
+    # substitute by online args and evaluate
+    eval(code, envir = env)
+
+    if (identical(ls(env), character(0))) {
+      warning("DDL code returned NULL. Returning empty  object")
+    }
+
+    # don't keep input further we don't want to keep input in the @env of teal_data
+    # but we want to keep other non-dataset objects created in the code
+    env_list <- as.list(env)
+    env_list <- env_list[names(env_list) != "input"]
```

Comment:
I'm not sure about this. Shouldn't all created variables be available for later use? Is it a valid assumption that they will never be used again?